In [1]:
import pandas as pd
import matplotlib.pyplot as plt


In [19]:
df = pd.read_csv('data/Labelled Yelp Dataset Edited.csv')
target = len(df.loc[df['Label'] == -1])
pd.set_option ('display.max_colwidth', None)            

In [42]:
print(f'''Number of targets: {target}
Percent of data target: {round(target / len(df),3)*100}
Number of rows: {len(df)}''')

Number of targets: 11154
Percent of data target: 10.0
Number of rows: 111798


There are 11154 fake reviews that make up 10 percent of the data. If we were to go ahead and classify everything as a real review, there would be a 90% accuracy rate. 

In [21]:
df.isna().sum()

User_id       0
Product_id    0
Rating        0
Date          0
Review        0
Label         0
dtype: int64

There are no null values in our data set.

In [40]:
df.loc[df['Label'] == -1].tail(20)

User_id  Product_id  Rating        Date  \
111331    69314          54       5    5/6/2013   
111332    69315          54       5   3/31/2013   
111333    69316          54       5    1/2/2013   
111334    69317          54       5   11/6/2011   
111335    69318          54       5  10/14/2011   
111644    69460          60       2   1/20/2014   
111645    69461          60       5   8/19/2014   
111646    69462          60       5   8/19/2014   
111647    69463          60       5   6/28/2014   
111648    69464          60       5  10/25/2013   
111649    69465          60       5  10/20/2013   
111743    69518         318       1   3/31/2013   
111744    69519         318       1  12/16/2012   
111745    69520         318       3  11/12/2012   
111746    69521         318       4   4/15/2013   
111747    69522         318       5   3/28/2014   
111748    69523         318       5   1/16/2014   
111749    69524         318       5  12/23/2013   
111750    69525         318       5    9/5/2013   
111751    69526         318       5  11/29/2012   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            Review  \
111331                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   The best restaraunt in New York City and one of the most historical places. I come here a lot. I love to support them instead of supporting other restaraunts in manhattan. It's my favorite, great food, great service, and most of all its a revolutionary war place which is my favorite! I can't believe George Washington slept, ate, and stayed here during the 1700s ! Bowling green is nearby which is a nice historical park, as well as other revolutionary war sites you should, visit. I live nearby on Chambers St and come here a lot to support them. Please support them the. Money goes to great cause. Museum is the best upstairs too.   
111332                                                                                                                                      

Here I look at how some of the fake or disingenuous reviews are worded. They all seem real to me, I must be missing something. I also wonder how or why they were rated genuine or not. Was it a person who did this? I could not find the answer on kaggle, I will assume the data is accurate and continue to go with it. 

After doing a little more research, I found out that Yelp also classifies them as 'recommended' or 'not recommended'. https://www.reviewtrackers.com/blog/fake-yelp-reviews/ is the website I discovered this at. Each post is carefully vetted by employees at Yelp. At the current state of that article(Jan 2019), 70% are recommended while 30% are not. Long rants are put in the non recommended, along with those that seem solicitated. They want all reviews to be genuine, without a restaraunt trying to boost their ratings in a disingenuous manner. Yelp also has more information such as which computers the reviews came from and activity of the users.

https://www.youtube.com/watch?v=KdG81QKdga0 is a link to a video where 5 star reviews are sold to businesses. There are Facebook groups where you can go to find people who will boost your score, with the 'elite' reviews being premium since they cannot be removed. This video is from September of 2020. This is why the fake reviews are so hard to spot. They are actual people writing these for money boosting the business rating. This leads to the question, can you also pay a group of people to bring down a competing business?

In [41]:
df.loc[df['Label'] == 1].tail(20)

User_id  Product_id  Rating        Date  \
111778    69545         318       5  12/20/2013   
111779    69546         318       5  11/21/2013   
111780    69547         318       5  11/20/2013   
111781    69548         318       2   11/1/2013   
111782    69549         318       5  10/12/2013   
111783    69550         318       5   10/6/2013   
111784    16438         318       3   9/15/2013   
111785    69551         318       5   9/12/2013   
111786    69552         318       5   8/16/2013   
111787    69553         318       5   8/12/2013   
111788    39956         318       5    8/8/2013   
111789    67537         318       4   6/12/2013   
111790     6579         318       4    6/8/2013   
111791    35030         318       4   4/21/2013   
111792    69554         318       5   4/20/2013   
111793    67541         318       2   3/29/2013   
111794    69555         318       4   2/13/2013   
111795    67575         318       5   11/9/2012   
111796    69556         318       5  10/13/2012   
111797    69557         318       3   9/13/2012   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            Review  \
111778                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            

Here I look at how the real reviews are worded. I find them worded similarly to the fake reviews so now I am certain I must be missing something. Hopefully the model can pick up on something I can't.